zilliqa/src/db.rs

```diff
@@ -221,6 +235,111 @@ impl Db {
         Ok(())
     }
 
+    pub fn create_checkpoint_path(&self, height: u64) -> Result<Option<Box<Path>>> {
+        let Some(base_path) = &self.path else {
+            // If we don't have on-disk persistency, disable checkpoints too
+            return Ok(None);
+        };
+        let snapshot_path = base_path.join("snapshots").join(height.to_string());
+        fs::create_dir_all(&snapshot_path)?;
+        Ok(Some(snapshot_path.into_boxed_path()))
+    }
+
+    pub fn load_trusted_checkpoint<P: AsRef<Path>>(
+        &self,
+        path: P,
+        hash: &Hash,
+        our_shard_id: u64,
+    ) -> Result<Block> {
+        let input = File::open(path)?;
+        let reader = BufReader::with_capacity(8192 * 1024, input); // 8 MiB read chunks
+        let trie_storage = Arc::new(self.state_trie()?); // TODO: what should we do if the trie DB already contains values?
+        let mut state_trie = EthTrie::new(trie_storage.clone());
+
+        let mut lines = reader.lines();
+        // decode blocks
+        let block = lines.next().ok_or(anyhow!(
+            "Invalid checkpoint file: missing block info on line 1"
+        ))??;
+        let block: Block = bincode::deserialize(&hex::decode(block.as_bytes())?)?;
+        if block.hash() != *hash {
+            return Err(anyhow!("Checkpoint does not match trusted hash"));
+        }
+        block.verify_hash()?;
+
+        let parent = lines.next().ok_or(anyhow!(
+            "Invalid checkpoint file: missing parent info on line 2"
+        ))??;
+        let parent: Block = bincode::deserialize(&hex::decode(parent.as_bytes())?)?;
+        parent.verify_hash()?;
+
+        if block.parent_hash() != parent.hash() {
+            return Err(anyhow!("Invalid checkpoint file: parent's blockhash does not correspond to checkpoint block"));
+        }
+
+        let shard_number: u64 = lines
+            .next()
+            .ok_or(anyhow!(
+                "Invalid checkpoint file: missing shard id on line 3"
+            ))??
+            .parse()?;
+        if shard_number != our_shard_id {
+            return Err(anyhow!("Invalid snapshot: chain ID mismatch. Snapshot ID: {shard_number}, our chain_id: {our_shard_id}"));
+        }
+
+        let our_max_view = self.get_latest_finalized_view()?.unwrap_or(0);
+        if block.view() <= our_max_view {
+            return Err(anyhow!("Snapshow is lower than our finalized view: snapshot is at {}, we are already at {}", block.view(), our_max_view));
+        }
+
+        // then decode state
+        for (idx, line) in lines.enumerate() {
+            let idx = idx + 2; // +1 because first line is just serialised block, +1 for 1-indexing
```

Comment:
Didn't we already read the serialised block above this?

Reply:
Yup, the `idx` is just a uint used for debugging/error messages when parsing. We pop off the first three lines, then start the loop so the iteration number = line number - 3.

However this is a holdover from when I only had the serialized block and nothing else - now I have it, the parent and the shard ID so it should actually be +4 here.